traffic_ops/app/db/reencrypt/reencrypt.go

```diff
@@ -0,0 +1,406 @@
+		fmt.Println("previousKey flag is required.")
+		os.Exit(0)
+	}
+	if newKeyLocation == nil || *newKeyLocation == "" {
+		fmt.Println("newKey flag is required.")
+		os.Exit(0)
+	}
+
+	newKey, err := readKey(*newKeyLocation)
+	if err != nil {
+		fmt.Println("reading newKey: ", err.Error())
+		os.Exit(0)
+	}
+
+	previousKey, err := readKey(*previousKeyLocation)
+	if err != nil {
+		fmt.Println("reading previousKey: ", err.Error())
+		os.Exit(0)
+	}
+
+	dbConfBytes, err := ioutil.ReadFile(PROPERTIES_FILE)
+	if err != nil {
+		fmt.Println("reading db conf '", PROPERTIES_FILE, "': ", err.Error())
+		os.Exit(0)
+	}
+
+	pgCfg := Config{}
+	err = json.Unmarshal(dbConfBytes, &pgCfg)
+	if err != nil {
+		fmt.Println("unmarshalling '", PROPERTIES_FILE, "': ", err.Error())
+		os.Exit(0)
+	}
+
+	sslStr := "require"
+	if !pgCfg.SSL {
+		sslStr = "disable"
+	}
+	db, err := sqlx.Open("postgres", fmt.Sprintf("postgres://%s:%s@%s:%d/%s?sslmode=%s&fallback_application_name=trafficvault", pgCfg.User, pgCfg.Password, pgCfg.Hostname, pgCfg.Port, pgCfg.DBName, sslStr))
+	if err != nil {
+		fmt.Println("opening database: " + err.Error())
+		os.Exit(0)
+	}
+
+	if err = reEncryptSslKeys(db, previousKey, newKey); err != nil {
+		fmt.Println("re-encrypting SSL Keys: ", err.Error())
+		os.Exit(0)
+	}
+	if err = reEncryptUrlSigKeys(db, previousKey, newKey); err != nil {
+		fmt.Println("re-encrypting URL Sig Keys: ", err.Error())
+		os.Exit(0)
+	}
+	if err = reEncryptUriSigningKeys(db, previousKey, newKey); err != nil {
+		fmt.Println("re-encrypting URI Signing Keys: ", err.Error())
+		os.Exit(0)
+	}
+	if err = reEncryptDNSSECKeys(db, previousKey, newKey); err != nil {
+		fmt.Println("re-encrypting DNSSEC Keys: ", err.Error())
+		os.Exit(0)
+	}
+
+	if err = updateKeyFile(previousKey, *previousKeyLocation, newKey); err != nil {
+		fmt.Println("updating the key file: ", err.Error())
+		os.Exit(0)
+	}
+
+	fmt.Println("Successfully re-encrypted keys for SSL Keys, URL Sig Keys, URI Signing Keys, and DNSSEC Keys.")
+}
+
+type Config struct {
+	DBName   string `json:"dbname"`
+	Hostname string `json:"hostname"`
+	User     string `json:"user"`
+	Password string `json:"password"`
+	Port     int    `json:"port"`
+	SSL      bool   `json:"ssl"`
+}
+
+func readKey(keyLocation string) ([]byte, error) {
+	var keyBase64 string
+	keyBase64Bytes, err := ioutil.ReadFile(keyLocation)
+	if err != nil {
+		return []byte{}, errors.New("reading file '" + keyLocation + "':" + err.Error())
+	}
+	keyBase64 = string(keyBase64Bytes)
+
+	key, err := base64.StdEncoding.DecodeString(keyBase64)
+	if err != nil {
+		return []byte{}, errors.New("AES key cannot be decoded from base64")
+	}
+
+	// verify the key works
+	_, err = aes.NewCipher(key)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return key, nil
+}
+
+func reEncryptSslKeys(db *sqlx.DB, previousKey []byte, newKey []byte) error {
+	tx, err := db.Begin()
+	if err != nil {
+		fmt.Println("transaction begin failed ", err, " ", tx)
+		os.Exit(0)
```

Comment:
Same as above RE: exit with failure code on failure

Reply:
Also, sometimes this function returns an error, and sometimes it forcibly exits the entire program - why not just always return the error and have the caller handle exiting or not?